### PR TITLE
fix: handle duplicate slugs

### LINKED ISSUE #132: [Events] Add unique slug conflict failover
The slug field for events is unique. It is generated based on the event title. So 2 events having the same title will have same slug. On creating the second event, the request will fail have because of the unique constraint.

Before creating an event, check how many events with the similar slug exists `LIKE '<slug>%'`. If count is non zero, append `-<count + 1>` to the new slug and create the event.

### PR DESCRIPTION
**Pull request by prabalsingh24**
## Issue
fixes: #132 
## Solution
count and modify the slug accordingly

## Screenshots (Desktop)
Screenshot_2020-06-09 Calendar CodingBlocks(1)

## Checklist
*PR will only be reviewed if the checklist is completed*
- [x] I have done extensive testing for my changes.
- [x] I have self-reviewed my PR and removed all un-necessary changes.
- [x] My changes works well on both desktop and mobile environments.
- [x] I have made sure that my changes solves the actual problem mentioned in the issue.
- [x] The PR contains only 1 commit and have been updated with `master`.

### CODE REVIEW

File: api/services/CalendarEvent/index.js [hunk 1]
```diff
@@ -8,9 +8,21 @@ import slugify from 'slugify';
 
 export default class CalendarEvent extends BaseModelService {
   @saveInstance
-  create(body) {
+  async create(body) {
```
Comment (0xnmn) [inline api/services/CalendarEvent/index.js]: Need to be inside a transaction. https://github.com/coding-blocks/gondor/blob/master/api/services/BaseModelService/index.js#L64

File: api/services/CalendarEvent/index.js [hunk 2]
```diff
@@ -8,9 +8,21 @@ import slugify from 'slugify';
...
+    const slug = slugify(body.title);
+    const numberOfSimilarSlugs = await Models.CalendarEvent.count({
+      where: {
+        slug: {
+          [Models.Sequelize.Op.startsWith]: slug,
+        },
+      },
+    });
```
Comment (0xnmn) [inline api/services/CalendarEvent/index.js:21]: add a new line.

File: api/services/CalendarEvent/index.js [hunk 3]
```diff
@@ -8,9 +8,21 @@ import slugify from 'slugify';
 
 export default class CalendarEvent extends BaseModelService {
   @saveInstance
-  create(body) {
+  async create(body) {
+    const slug = slugify(body.title);
+    const numberOfSimilarSlugs = await Models.CalendarEvent.count({
```
Comment (0xnmn) [inline api/services/CalendarEvent/index.js]: `similarSlugsCount`

File: api/services/CalendarEvent/index.js [hunk 4]
```diff
@@ -8,9 +8,21 @@ import slugify from 'slugify';
...
+      where: {
+        slug: {
+          [Models.Sequelize.Op.startsWith]: slug,
+        },
+      },
+    });
+    const updatedSlug = numberOfSimilarSlugs
```
Comment (0xnmn) [inline api/services/CalendarEvent/index.js]: simply do not declare slug as const and override it.

**Review by 0xnmn — CHANGES_REQUESTED** [review]

File: api/services/CalendarEvent/index.js (hunk 1 above)
Comment (prabalsingh24) [reply api/services/CalendarEvent/index.js]: Umm I don't get it.
Do you mean like this?

```
  @saveInstance
  @withTransaction
  async create(body, transaction) {
    let slug = slugify(body.title);
    const similarSlugsCount = await Models.CalendarEvent.count({
      where: {
        slug: {
          [Models.Sequelize.Op.startsWith]: slug,
        },
      },
      transaction,
    });
    .....
```

Comment (0xnmn) [reply api/services/CalendarEvent/index.js]: yes, learn more here: https://sequelize.org/v5/class/lib/transaction.js~Transaction.html

**Comment by prabalsingh24** [comment]
Made these changes :)

File: api/services/CalendarEvent/index.js [hunk 5]
```diff
@@ -8,9 +8,24 @@ import slugify from 'slugify';
 
 export default class CalendarEvent extends BaseModelService {
   @saveInstance
-  create(body) {
+  @withTransaction
```
Comment (0xnmn) [inline api/services/CalendarEvent/index.js:11]: What is the purpose of the transaction when you are not passing it to the create query.

**Review by 0xnmn — CHANGES_REQUESTED** [review]

File: api/models/calendar_event.js [hunk 6]
```diff
@@ -61,6 +61,7 @@ module.exports = (sequelize, DataTypes) => {
       slug: {
         type: DataTypes.STRING,
         allowNull: true,
+        unique: true,
```
Comment (0xnmn) [inline api/models/calendar_event.js:64]: allowNull also should be false.

**Review by 0xnmn — CHANGES_REQUESTED** [review]

File: api/models/calendar_event.js (hunk 6 above)
Comment (prabalsingh24) [reply api/models/calendar_event.js:64]: Fixed it :)

Comment (prabalsingh24) [reply api/models/calendar_event.js:64]: fixed it :)

**Comment by prabalsingh24** [comment]
Ping @hereisnaman

**Review by 0xnmn — APPROVED** [review]

**Pull request merged**